商品購入機能

kenshiro-wata opened the pull request:
WHAT
1.  pay.jpを使い、仮のクレジットカードで商品の購入を行えるようにしました。
2. formオブジェクトを使い、購入した際に購入管理テーブルと購入者情報を同時に更新することを可能にしました。
3. 商品が購入されたあとはSOLDの文字が画像に貼られ、redirect_toメソッドを使うことで2度同じ商品が購入されることを防いでいます。

WHY
1. ECサイトでの商品購入にはクレジットカードの決済が必要なため、pay.jpを使いました。
2. 商品は購入した後取引を行うため、商品情報と購入したユーザー情報を１つのテーブルで管理する必要があります。そのテーブルと商品購入情報を管理するテーブルを商品購入の際に同時に保存をしたいのでformオブジェクトを使用しました。

https://gyazo.com/5b98f0ea9ed418f51ed4b41537be438b
https://gyazo.com/0bbaa0598c43ffc54e3d403a70e46a50
https://gyazo.com/2e472579ef7af7f9cf47d214be829dde
https://gyazo.com/e776753658ba5a3e8510a2d3521104c8

よろしくお願いします！

kenshiro-wata committed:
pay.jp の導入

kenshiro-wata committed:
購入画面を表示させる実装

kenshiro-wata committed:
クレジット決済処理の実装

kenshiro-wata committed:
manage モデルの作成

kenshiro-wata committed:
orderモデルの作成

kenshiro-wata committed:
各モデルファイルにアソシエーションを記述

kenshiro-wata committed:
商品詳細ページの修正

kenshiro-wata committed:
formオブジェクトの実装

kenshiro-wata committed:
購入商品情報の表示

kenshiro-wata committed:
購入商品データの保存の実装

kenshiro-wata committed:
決済処理の環境変数の設定

kenshiro-wata committed:
エラーメッセージの表示

kenshiro-wata committed:
SOLD OUT済み商品の表示

kenshiro-wata committed:
購入済商品の再購入を防ぐ実装

- 出品者はURLを直接入力して購入ページに遷移しようとすると、トップページに遷移すること。
- URLを直接入力して購入済み商品の購入ページへ遷移しようとすると、トップページに遷移すること

kenshiro-wata committed:
モデル単体テストコード

このアプリケーションでは、ManageモデルとOrderモデルを作成しています。しかし、それらはFormオブジェクトであるManageOrderクラスでバリデーションを判断しているため、ManageOrderクラスについてのテストコードを記述しています。

kenshiro-wata committed:
Rubocop

Katsuno0814 commented on app/forms/manage_order.rb:
```diff
@@ -0,0 +1,21 @@
+class ManageOrder
+  include ActiveModel::Model
+  # attr_accessor :name, :email, :password, :password_confirmation, :last_name, :first_name, :last_name_kana, :first_name_kana, :birth, :image, :name, :description, :category_id, :status_id, :shipping_charges_id, :address_id, :date_takes_id, :price, :postal_code, :prefecture, :address_city, :block_number, :building_number, :phone_number, :manage
```
こちら不要なコメントアウトは削除いたしましょう！！

Katsuno0814 commented on app/javascript/packs/application.js:
```diff
@@ -4,11 +4,12 @@
 
 require("@rails/ujs").start()
-require("turbolinks").start()
+// require("turbolinks").start()
```
必要でなければ不要なコメントアウトは削除いたしましょう！

Katsuno0814 commented on app/views/items/show.html.erb:
```diff
@@ -28,9 +30,9 @@
       <%= link_to '削除', item_path(@item.id), method: :delete, class:'item-destroy' %>
-    <% elsif user_signed_in? && current_user.id != @item.user_id%>
+    <% elsif user_signed_in? && current_user.id != @item.user_id && @item.manage.blank?%>
       <%# 商品が売れていない場合はこちらを表示しましょう %>
```
<%# 商品が売れていない場合はこちらを表示しましょう %>
こちらも不要でしたら削除いたしましょう！

Katsuno0814 commented on spec/models/manage_order_spec.rb:
```diff
@@ -0,0 +1,62 @@
+      @manage_order = FactoryBot.build(:manage_order)
+    end
+
+    it 'postal_code, prefecture, address_city, block_number, phone_number, tokenが存在すれば登録できること' do
+      expect(@manage_order).to be_valid
+    end
```
こちらcontextを使い正常系と異常系で分けましょう！

Katsuno0814 commented on spec/models/manage_order_spec.rb:
```diff
@@ -0,0 +1,62 @@
+      expect(@manage_order.errors.full_messages).to include('Postal code is invalid')
+    end
+
+    it '電話番号は１１桁以内でないと登録できないこと' do
```
電話番号にハイフンがない場合登録できないことを追加いたしましょう！

Katsuno0814 reviewed (commented):
レビュー依頼ありがとうございます！
こちらのプルリクエストですが、プルリクエストにWhatとWhyが書かれていません。
What（どのような実装をしているのか）と Why（なぜこの実装が必要なのか）を書くことで、
このブランチは何を実装していて、なぜその実装を行ったのかが他の人に伝えることができます。
What をしっかり書くことで、他者にコードを見てもらう際に、スムーズな理解を促すことができ、Whyを書くことでなぜこの実装を行ったのか、本来の目的に沿ったレビュアーとのコミュニケーションが可能となります。
お手数かけますが、記入後再度提出いただけますよう、よろしくお願いします。
https://master.tech-camp.in/curriculums/4675

kenshiro-wata committed:
修正１

- 不要なコメントアウトの削除

- テストコードにてcontextを使い正常系と異常系で分ける

- テストコードにて、電話番号にハイフンがある場合登録できないことを追加

isyt729 commented:
レビュー依頼ありがとうございます１
LGTMといたします！
引き続き実装頑張っていきましょう！

kenshiro-wata commented:
ありがとうございました！！
